### ISSUE TITLE
False positive on collapsible_if with if let

### ISSUE THREAD

**Issue by llogiq** [rust-lang/rust-clippy#1197]
This is from [linefeed](https://github.com/murarth/linefeed/blob/master/src/reader.rs#L2114-L2124):

```
fn read_init(&mut self) {
    if let Some(path) = env_init_file() {
        // If `INPUTRC` is present, even if invalid, parse nothing else.
        // Thus, an empty `INPUTRC` will inhibit loading configuration.
        self.read_init_file_if_exists(Some(path));
    } else {
        if !self.read_init_file_if_exists(user_init_file()) {
            self.read_init_file_if_exists(system_init_file());
        }
    }
}
```

If you try to collapse this, rustc won't have it:

```
error[E0301]: cannot mutably borrow in a pattern guard
    --> src/reader.rs:2116:20
     |
2116 |         } else if !self.read_init_file_if_exists(user_init_file()) {
     |                    ^^^^ borrowed mutably in pattern guard

error: aborting due to previous error
```

The difference appears to be that the `else if` part of an `if let` desugars to a guard. And guards appear to have different borrowing rules than simple `if` expressions (I'm not sure why). Perhaps that is a bug in borrowck, but as long as it is there, we should at least note it in the docs.

**Comment by mcarton** [comment]
Commits that seems to add `if let` desugaring: https://github.com/rust-lang/rust/commit/0e6ff432dc83e5675163b1c29ce73dbc5e18ba24#diff-7fde1475846894e59317d75947bdb921R72.

```rust
+        // Desugar ExprIfLet
+        // From: `if let <pat> = <expr> <body> [<elseopt>]`
+        // to:
+        //
+        //   match <expr> {
+        //     <pat> => <body>,
+        //     [_ if <elseopt_if_cond> => <elseopt_if_body>,]
+        //     _ => [<elseopt> | ()]
+        //   }
```
There isn't much explanation about why this is done that way. But it certainly seems surprising because it leads to very unintuitive errors. I think we should report that on rustc's issue tracker.
See also https://github.com/rust-lang/rust/issues/28449.

**Comment by Arnavion** [comment]
https://github.com/rust-lang/rust/issues/28449 is closed. https://github.com/rust-lang/rust/issues/37510 is another issue for it that's still open.

**Comment by camsteffen** [comment]
This must be fixed by now, especially with NLL.
